core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Resources;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Properties;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static java.lang.Integer.parseInt;
+import static java.lang.String.format;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSingleVersionBackwardCompatibility
+{
+    @Test
+    public void test()
+            throws Exception
+    {
+        assertThat(getCurrentVersionEntities()).containsAll(getPreviousVersionEntities());
+    }
+
+    private static List<String> getCurrentVersionEntities()
+            throws IOException
+    {
+        return getSpiEntities(getSystemClassLoader(), true);
+    }
+
+    private static List<String> getPreviousVersionEntities()
+            throws Exception
+    {
+        String previousVersion = Integer.toString(parseInt(getCurrentVersion().replace("-SNAPSHOT", "")) - 1);
+        URL slice = download(format("https://repo1.maven.org/maven2/io/airlift/slice/%1$s/slice-%1$s.jar", getSliceVersion()));
+        URL spi = download(format("https://repo1.maven.org/maven2/io/trino/trino-spi/%1$s/trino-spi-%1$s.jar", previousVersion));
```

Comment:
This will cause tests to fail in offline mode, or in any environment that doesn't have direct access to maven central. In some organizations, all access is performed through a proxy or a local maven repository.

Reply:
So it looks like I need to download artifacts via maven, so it should use local maven repository in most of the cases.

Reply:
You could probably use https://maven.apache.org/plugins/maven-dependency-plugin/copy-mojo.html#dependency-copy

Reply:
I think it's reasonable to require internet access during tests.
For example, docker-based tests won't pass if it is not possible to download the container images.

cc @nineinchnick

Reply:
I agree using local maven repo would be better though. Eg it would cover for retries required to cope with central's instability.

This can help

```xml
        <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <executions>
                    <execution>
                        <id>copy</id>
                        <phase>package</phase>
                        <goals>
                            <goal>copy</goal>
                        </goals>
                        <configuration>
                            <skip>false</skip>
                            <artifactItems>
                                <artifactItem>
                                    <groupId>io.trino</groupId>
                                    <artifactId>trino-cli</artifactId>
                                    <version>${dep.trino.version}</version>
                                    <classifier>executable</classifier>
                                    <type>jar</type>
                                    <outputDirectory>${project.build.directory}</outputDirectory>
                                </artifactItem>
```

Reply:
> docker-based tests won't pass if it is not possible to download the container images.

You can always pre-download them and run the tests offline.